File: lib/py/uapi/internal/validation/ValidateUnion.py
from typing import TYPE_CHECKING, cast
from uapi.internal.types.UUnion import _UNION_NAME
from uapi.internal.validation.ValidationFailure import ValidationFailure

if TYPE_CHECKING:
    from uapi.internal.validation.ValidateContext import ValidateContext
    from uapi.internal.types.UStruct import UStruct
    from uapi.internal.types.UTypeDeclaration import UTypeDeclaration


def validate_union(value: object,
                   name: str, tags: dict[str, 'UStruct'], ctx: 'ValidateContext') -> list['ValidationFailure']:
    from uapi.internal.validation.GetTypeUnexpectedValidationFailure import get_type_unexpected_validation_failure
    from uapi.internal.validation.ValidateUnionTags import validate_union_tags

    if isinstance(value, dict):
        selected_tags: dict[str, object]
        if name.startswith("fn."):
            selected_tags = {name: ctx.select.get(
                name) if ctx.select else None}
        else:
            selected_tags = cast(
                dict[str, object], ctx.select.get(name) if ctx.select else None)
        return validate_union_tags(tags, selected_tags, value, ctx)
    else:
        return get_type_unexpected_validation_failure([], value, _UNION_NAME)

File: lib/java/src/main/java/uapi/internal/validation/ValidateUnion.java
package uapi.internal.validation;

import static uapi.internal.types.UUnion._UNION_NAME;
import static uapi.internal.validation.GetTypeUnexpectedValidationFailure.getTypeUnexpectedValidationFailure;
import static uapi.internal.validation.ValidateUnionTags.validateUnionTags;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import uapi.internal.types.UStruct;

public class ValidateUnion {
    public static List<ValidationFailure> validateUnion(Object value,
            String name, Map<String, UStruct> tags, ValidateContext ctx) {
        if (value instanceof Map<?, ?> m) {
            Map<String, Object> selectedTags;
            if (name.startsWith("fn.")) {
                selectedTags = new HashMap<>();
                selectedTags.put(name, ctx.select == null ? null : ctx.select.get(name));
            } else {
                selectedTags = ctx.select == null ? null : (Map<String, Object>) ctx.select.get(name);
            }
            return validateUnionTags(tags, selectedTags, m, ctx);
        } else {
            return getTypeUnexpectedValidationFailure(List.of(), value, _UNION_NAME);
        }
    }
}
